### ISSUE TITLE
New Empty Stack USFM bombing issue

### ISSUE THREAD

**Issue by johnml1135** [sillsdev/serval#388]
```
	
machine-job <fail> fail: SIL.Machine.AspNetCore.Services.NmtPreprocessBuildJob[0]
      Build faulted (66401d23d874d81aa4ecbfa8)
      System.InvalidOperationException: Stack empty.
         at System.Collections.Generic.Stack`1.ThrowForEmptyStack()
         at System.Collections.Generic.Stack`1.Pop()
         at SIL.Machine.Corpora.ScriptureRefUsfmParserHandlerBase.NextElement(String marker) in /app/src/SIL.Machine/Corpora/ScriptureRefUsfmParserHandlerBase.cs:line 219
         at SIL.Machine.Corpora.ScriptureRefUsfmParserHandlerBase.StartParentElement(String marker) in /app/src/SIL.Machine/Corpora/ScriptureRefUsfmParserHandlerBase.cs:line 225
         at SIL.Machine.Corpora.UsfmTextBase.TextRowCollector.StartPara(UsfmParserState state, String marker, Boolean unknown, IReadOnlyList`1 attributes) in /app/src/SIL.Machine/Corpora/UsfmTextBase.cs:line 91
         at SIL.Machine.Corpora.UsfmParser.ProcessToken()
         at SIL.Machine.Corpora.UsfmTextBase.GetVersesInDocOrder() in /app/src/SIL.Machine/Corpora/UsfmTextBase.cs:line 34
         at SIL.Machine.Corpora.ScriptureText.GetRows() in /app/src/SIL.Machine/Corpora/ScriptureText.cs:line 21
         at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
         at SIL.Machine.Corpora.ParallelTextCorpus.TargetCorpusEnumerator.CollectVerses() in /app/src/SIL.Machine/Corpora/ParallelTextCorpus.cs:line 596
         at SIL.Machine.Corpora.ParallelTextCorpus.TargetCorpusEnumerator.MoveNext() in /app/src/SIL.Machine/Corpora/ParallelTextCorpus.cs:line 568
         at SIL.Machine.Corpora.ParallelTextCorpus.GetRows()+MoveNext() in /app/src/SIL.Machine/Corpora/ParallelTextCorpus.cs:line 289
         at SIL.Machine.Corpora.CorporaExtensions.ExtractScripture(ITextCorpus corpus, ITextCorpus refCorpus)+MoveNext() in /app/src/SIL.Machine/Corpora/CorporaExtensions.cs:line 285
         at System.Linq.Enumerable.ZipIterator[TFirst,TSecond,TResult](IEnumerable`1 first, IEnumerable`1 second, Func`3 resultSelector)+MoveNext()
         at SI
```

**Comment by ddaspit** [comment]
It would be good to get the PT project that caused this error.

**Comment by pmachapman** [comment]
@johnml1135 I have another project that causes this to happen on SIL.Machine `master`, which I have emailed to you and Damien. The stack crash occurs on QA/Prod (with an issue that is fixed by #198), but when the latest master is used, another empty stack issue occurs in Nahum in the target project I sent through via email, so it may be a regression caused by that fix?

**Stack Trace**
```
System.Collections.dll!System.Collections.Generic.Stack<System.Text.StringBuilder>.ThrowForEmptyStack()
System.Collections.dll!System.Collections.Generic.Stack<System.Text.StringBuilder>.Pop()
SIL.Machine.dll!SIL.Machine.Corpora.UsfmTextBase.TextRowCollector.EndNonVerseText(SIL.Machine.Corpora.UsfmParserState state, SIL.Machine.Corpora.ScriptureRef scriptureRef) Line 243
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\UsfmTextBase.cs(243)
SIL.Machine.dll!SIL.Machine.Corpora.ScriptureRefUsfmParserHandlerBase.EndNonVerseText(SIL.Machine.Corpora.UsfmParserState state) Line 190
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\ScriptureRefUsfmParserHandlerBase.cs(190)
SIL.Machine.dll!SIL.Machine.Corpora.ScriptureRefUsfmParserHandlerBase.EndPara(SIL.Machine.Corpora.UsfmParserState state, string marker) Line 98
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\ScriptureRefUsfmParserHandlerBase.cs(98)
SIL.Machine.dll!SIL.Machine.Corpora.UsfmParser.CloseElement(bool closed) Line 632
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\UsfmParser.cs(632)
SIL.Machine.dll!SIL.Machine.Corpora.UsfmParser.ProcessToken() Line 216
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\UsfmParser.cs(216)
SIL.Machine.dll!SIL.Machine.Corpora.UsfmParser.ProcessTokens() Line 132
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\UsfmParser.cs(132)
SIL.Machine.dll!SIL.Machine.Corpora.UsfmParser.Parse(string usfm, SIL.Machine.Corpora.IUsfmParserHandler handler, SIL.Machine.Corpora.UsfmStylesheet stylesheet, SIL.Scripture.ScrVers versification, bool preserveWhitespace) Line 44
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\UsfmParser.cs(44)
SIL.Machine.dll!SIL.Machine.Corpora.UsfmTextBase.GetVersesInDocOrder() Line 36
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\UsfmTextBase.cs(36)
SIL.Machine.dll!SIL.Machine.Corpora.ScriptureText.GetRows() Line 21
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\ScriptureText.cs(21)
SIL.Machine.dll!SIL.Machine.Corpora.DictionaryTextCorpus.GetRows.AnonymousMethod__21_1(SIL.Machine.Corpora.IText t) Line 50
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\DictionaryTextCorpus.cs(50)
System.Linq.dll!System.Linq.Enumerable.SelectManySingleSelectorIterator<SIL.Machine.Corpora.IText, SIL.Machine.Corpora.TextRow>.MoveNext()
SIL.Machine.dll!SIL.Machine.Corpora.ParallelTextCorpus.TargetCorpusEnumerator.MoveNext() Line 578
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\ParallelTextCorpus.cs(578)
SIL.Machine.dll!SIL.Machine.Corpora.ParallelTextCorpus.GetRows() Line 289
	at D:\Source\Repos\machine\src\SIL.Machine\Corpora\ParallelTextCorpus.cs(289)
System.Private.CoreLib.dll!System.Collections.Generic.List<SIL.Machine.Corpora.ParallelTextRow>.List(System.Collections.Generic.IEnumerable<SIL.Machine.Corpora.ParallelTextRow> collection)
System.Linq.dll!System.Linq.Enumerable.ToList<SIL.Machine.Corpora.ParallelTextRow>(System.Collections.Generic.IEnumerable<SIL.Machine.Corpora.ParallelTextRow> source)
SIL.Machine.Tests.dll!SIL.Machine.Corpora.UsfmManualTests.ParseParallelCorpus() Line 35
	at D:\Source\Repos\machine\tests\SIL.Machine.Tests\Corpora\UsfmManualTests.cs(35)
[Native to Managed Transition]
[Managed to Native Transition]
System.Private.CoreLib.dll!System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(object obj, System.Reflection.BindingFlags invokeAttr)
nunit.framework.dll!NUnit.Framework.Internal.Reflect.InvokeMethod(System.Reflection.MethodInfo method, object fixture, object[] args) Line 222
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Reflect.cs(222)
nunit.framework.dll!NUnit.Framework.Internal.MethodWrapper.Invoke(object fixture, object[] args) Line 186
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\MethodWrapper.cs(186)
nunit.framework.dll!NUnit.Framework.Internal.Commands.TestMethodCommand.InvokeTestMethod(NUnit.Framework.Internal.TestExecutionContext context) Line 92
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs(92)
nunit.framework.dll!NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(NUnit.Framework.Internal.TestExecutionContext context) Line 87
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs(87)
nunit.framework.dll!NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(NUnit.Framework.Internal.TestExecutionContext context) Line 65
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Commands\TestMethodCommand.cs(65)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork.AnonymousMethod__0() Line 79
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItem.cs(79)
nunit.framework.dll!NUnit.Framework.Internal.ContextUtils.DoIsolated.AnonymousMethod__0(object _) Line 40
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\ContextUtils.cs(40)
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)
nunit.framework.dll!NUnit.Framework.Internal.ContextUtils.DoIsolated(System.Threading.ContextCallback callback, object state) Line 55
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\ContextUtils.cs(55)
nunit.framework.dll!NUnit.Framework.Internal.ContextUtils.DoIsolated<NUnit.Framework.Internal.TestResult>(System.Func<NUnit.Framework.Internal.TestResult> func) Line 40
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\ContextUtils.cs(40)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItem.PerformWork() Line 79
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItem.cs(79)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() Line 494
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(494)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.Execute() Line 239
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(239)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem work) Line 84
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItemDispatcher.cs(84)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() Line 298
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(298)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork() Line 110
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(110)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() Line 494
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(494)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.Execute() Line 239
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(239)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem work) Line 84
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItemDispatcher.cs(84)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() Line 298
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(298)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork() Line 110
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(110)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() Line 494
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(494)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.Execute() Line 239
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(239)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem work) Line 84
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItemDispatcher.cs(84)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() Line 298
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(298)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork() Line 110
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(110)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() Line 494
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(494)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.Execute() Line 239
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(239)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem work) Line 84
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItemDispatcher.cs(84)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() Line 298
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(298)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork() Line 110
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(110)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() Line 494
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(494)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.Execute() Line 239
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(239)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.Dispatch(NUnit.Framework.Internal.Execution.WorkItem work) Line 84
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItemDispatcher.cs(84)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.RunChildren() Line 298
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(298)
nunit.framework.dll!NUnit.Framework.Internal.Execution.CompositeWorkItem.PerformWork() Line 110
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\CompositeWorkItem.cs(110)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.RunOnCurrentThread() Line 494
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(494)
nunit.framework.dll!NUnit.Framework.Internal.Execution.WorkItem.Execute() Line 239
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\WorkItem.cs(239)
nunit.framework.dll!NUnit.Framework.Internal.Execution.SimpleWorkItemDispatcher.RunnerThreadProc() Line 90
	at C:\Src\NUnit\nunit\src\NUnitFramework\framework\Internal\Execution\SimpleWorkItemDispatcher.cs(90)
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)

```
